### ISSUE TITLE
Problems with tests

### ISSUE THREAD

**Issue by frenzymadness** [haypo/faulthandler#32]
Hello.

I am updating python2-faulthandler in Fedora to version 3.0 and I found two issues with tests.

First one is caused by different default name of Python threads which causes a problem with regular expression matching like:
```
======================================================================
FAIL: test_gil_released (tests.FaultHandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/faulthandler-3.0/tests.py", line 349, in test_gil_released
    'Segmentation fault')
  File "/builddir/build/BUILD/faulthandler-3.0/tests.py", line 214, in check_fatal_error
    self.check_error(code, line_number, fatal_error, **kw)
  File "/builddir/build/BUILD/faulthandler-3.0/tests.py", line 209, in check_error
    self.assertRegex(output, regex)
  File "/builddir/build/BUILD/faulthandler-3.0/tests.py", line 829, in assertRegex
    raise self.failureException(msg)
AssertionError: Regex didn't match: '^Fatal Python error: Segmentation fault\n\nCurrent\\ thread\\ XXX\\ \\<python\\>\\ \\(most\\ recent\\ call\\ first\\):\n  File "<string>", line 3 in <module>' not found in u'Fatal Python error: Segmentation fault\n\nCurrent thread XXX <python2> (most recent call first):\n  File "<string>", line 3 in <module>'
```
I can fix this one with just one `sed` before text execution and I don't know if this change is applicable for all platforms/OS so I haven't prepared PR yet.

The second one is harder to crack. Failing test is
```
======================================================================
FAIL: test_enable_env_var (tests.FaultHandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/faulthandler-3.0/tests.py", line 379, in test_enable_env_var
    self.assertEqual(output, ["True"])
AssertionError: Lists differ: [u'False'] != ['True']

First differing element 0:
u'False'
'True'

- [u'False']
+ ['True']
```
And this fail looks reasonable because when I try it manually it seems that enabling faulthandler with environment variable is not working.
```
# ls -1
faulthandler-3.0-py2.7.egg-info
faulthandler.pth
faulthandler.so
# PYTHONFAULTHANDLER='x' python
Python 2.7.13 (default, Sep  1 2017, 16:13:49) 
[GCC 7.2.1 20170829 (Red Hat 7.2.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import faulthandler, os
>>> os.environ.get('PYTHONFAULTHANDLER')
'x'
>>> faulthandler.is_enabled()
False
```
Could you please help me to fix at least the second one?

**Comment by haypo** [comment]
> FAIL: test_gil_released (tests.FaultHandlerTests)

Can you please check if the commit 67b661ebdbbf4067bd173af4d61ba815b9d49616 fixed this test?

> FAIL: test_enable_env_var (tests.FaultHandlerTests)

This test requires that the .pth file is installed. How do you run tests? Did you install faulthandler?

faulthandler.pth is the current directory isn't enough.

I'm running tests with "tox" which compiles faulthandler, installs the C extension but also installs the .pth file.

**Comment by frenzymadness** [comment]
> Can you please check if the commit 67b661e fixed this test?

It works, thanks. I'll include this patch in the RPM sources until you release a new version with this change included.

> This test requires that the .pth file is installed. How do you run tests? Did you install faulthandler?

Faulthandlers files are installed during a build of RPM package into buildroot directory and I am using this directory as PYTHONPATH environment variable during tests execution.

```PYTHONPATH=/builddir/build/BUILDROOT/python2-faulthandler-3.0-1.fc28.x86_64/usr/lib64/python2.7/site-packages nosetests-2.7 --exclude=test_stack_overflow```

PTH file is also installed in the same folder

```copying build/lib.linux-x86_64-2.7/faulthandler.pth -> /builddir/build/BUILDROOT/python2-faulthandler-3.0-1.fc28.x86_64/usr/lib64/python2.7/site-packages```

**Comment by haypo** [comment]
Sorry, I have no clue for the pth file, I didn't write it. It was written by @ionelmc.

**Comment by ionelmc** [comment]
Technically setting PYTHONPATH to something doesn't make it a site, it needs explicit activation (see https://docs.python.org/2/library/site.html#site.addsitedir).

Perhaps there a different, and more correct way to run the tests? @frenzymadness, how can I reproduce this locally?

**Comment by frenzymadness** [comment]
It's hard to explain but I'll try it. The build of RPM package contains a few phases:
- Preparation - unpacking sources, applying patches etc.
- Build - Basically `python setup.py build` for python packages
- Install - Install files from build directory to buildroot directory which is something like root for RPM package - `python setup.py install` for python packages apparently
- Check - run tests on installed files
- Files - pack files in buildroot to the new RPM package

The basic approach to enable import installed package during tests is based on changing PYTHONPATH to contains BUILDROOT folder with newly installed files. But as you mentioned, setting PYTHONPATH is not enough to process .pth files.

I think that it is reproducible locally by building faulthandler without installing them and then running tests against build binaries in `/build/` folder. Basically, you want to run tests which need to import faulthandler from a directory which is not a site-packages directory and you also need .pth file to be processed.

**Comment by frenzymadness** [comment]
Ok, I fixed it.

Because I need to process BUILDROOT directory as standard site directory I create `sitecustomize.py` file in BUILDROOT directory with this content:
```
import site
site.addsitedir("/builddir/build/BUILDROOT/python2-faulthandler-3.0-1.fc28.x86_64/usr/lib64/python2.7/site-packages")
```
And then the .pth file is processed during import of faulthandler. I also have to remove `sitecustomize.py` after tests apparently.

Thank you for your help. I again learned something new.

**Comment by haypo** [comment]
>> Can you please check if the commit 67b661e fixed this test?

> It works, thanks. I'll include this patch in the RPM sources until you release a new version with this change included.

Do you want a release with the fix?

**Comment by frenzymadness** [comment]
> Do you want a release with the fix?

It's not necessary. I included mentioned patch to RPM so I can wait until you'll release new version with this patch included.

**Comment by stefano-m** [comment]
@vstinner 
NixOS has the same problem https://github.com/NixOS/nixpkgs/issues/34415

But the patch would not work because NixOS is stricter and uses `python2.7` not just `python2`.

Any chance this can be included?

Thanks for your time.

**Comment by kapsh** [comment]
Same fail on Exherbo where package manager runs test specifying full name of Python binary. Would you kindly to change regular expression to somehting like [this](https://regex101.com/r/DWtH2Y/3) and post an updated release?

**Comment by dotlambda** [comment]
There is already a PR for this: https://github.com/vstinner/faulthandler/pull/34

**Comment by kapsh** [comment]
Thank you @dotlambda , honestly speaking I didn't search PRs before posting.

**Comment by vstinner** [comment]
> There is already a PR for this: #34

Oh, I forgot that one. I merged the PR. Tell me if you need a release.

**Comment by kapsh** [comment]
Tag would be great, yes.

**Comment by vstinner** [comment]
I released faulthandler 3.1. Building wheel packages on Windows (32 and 64 bits binaries for Python 2.7) is always annoying, but it's now done. Enjoy.
